From 2e77e028609a3273fca02c324acf91c2dd6ddbf6 Mon Sep 17 00:00:00 2001
From: Stephan Mueller <smueller@chronox.de>
Date: Tue, 3 Jan 2023 11:21:52 +0100
Subject: [PATCH 29/36] lrng: fix unlocking of backed entropy

The lock disables interrupts, so we must disable them also.

Reported-by: Oleksandr Natalenko
Signed-off-by: Stephan Mueller <smueller@chronox.de>
Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 drivers/char/lrng/lrng_interface_aux.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/char/lrng/lrng_interface_aux.c b/drivers/char/lrng/lrng_interface_aux.c
index b42928305..718fb4574 100644
--- a/drivers/char/lrng/lrng_interface_aux.c
+++ b/drivers/char/lrng/lrng_interface_aux.c
@@ -176,17 +176,17 @@ void invalidate_batched_entropy(void)
 		batched_entropy = per_cpu_ptr(&batched_entropy_u8, cpu);
 		spin_lock_irqsave(&batched_entropy->batch_lock, flags);
 		batched_entropy->position = 0;
-		spin_unlock(&batched_entropy->batch_lock);
+		spin_unlock_irqrestore(&batched_entropy->batch_lock, flags);
 
 		batched_entropy = per_cpu_ptr(&batched_entropy_u16, cpu);
 		spin_lock_irqsave(&batched_entropy->batch_lock, flags);
 		batched_entropy->position = 0;
-		spin_unlock(&batched_entropy->batch_lock);
+		spin_unlock_irqrestore(&batched_entropy->batch_lock, flags);
 
 		batched_entropy = per_cpu_ptr(&batched_entropy_u32, cpu);
 		spin_lock_irqsave(&batched_entropy->batch_lock, flags);
 		batched_entropy->position = 0;
-		spin_unlock(&batched_entropy->batch_lock);
+		spin_unlock_irqrestore(&batched_entropy->batch_lock, flags);
 
 		batched_entropy = per_cpu_ptr(&batched_entropy_u64, cpu);
 		spin_lock(&batched_entropy->batch_lock);
-- 
2.39.0.rc2.1.gbd5df96b79

